src/execute.rs

```diff
@@ -44,8 +48,14 @@ pub fn execute_create_offer(
     // TODO: Consider a different order of checks: Now, you might get a not approved error, after which you approved, but actually there is another error, like the peer is not the right owner.
     //          Then you've approved the contract but no offer has been made, which feels a bit unsafe.
     for token in offered_tokens.clone() {
+        // TODO: don't need both to check owner and do an approval, because if the approval works, then you are already the owner
+        // Maybe only check only_owner..
+
         // TODO: [OPTIMISATION] See if we can levarage the OwnerOfResponse.Approvals for checking if the contract has been approved
-        let _ = only_owner(deps.as_ref(), &info, &token.collection, token.token_id)?;
+        only_owner(deps.as_ref(), &info, &token.collection, token.token_id)?;
+
+        // TODO: this approval check should happen on transfer, not on create offer
+        // needs to happen before final execution
```

Comment:
the may concern here by checking both is querying also takes gas, but I guess if you introduce a bundle limit you can still check both.

Maybe just checking only_owner approvals will be checked again when the offer is executed.

Reply:
Okay so the NFTs are transferred when the peer accepts the offer, which means we already need to have an approval from the sender at that point. The only way to do this is to gather the approval when the offer is created.

Reply:
I know this but this can be done at the front end level and checked front end side, if that makes sense. See my previous comment of how much gas is taking to make an offer.